src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -5909,7 +5910,14 @@ Result<ASR::TranslationUnit_t*> python_ast_to_asr(Allocator &al, LocationManager
         return res.error;
     }
     ASR::TranslationUnit_t *tu = ASR::down_cast2<ASR::TranslationUnit_t>(unit);
-    LFORTRAN_ASSERT(asr_verify(*tu, true, diagnostics));
+    PassUtils::UpdateDependenciesVisitor symbol_table_update(al);
+    symbol_table_update.visit_TranslationUnit(*tu);
```

Comment:
I think that this is a problematic design for two reasons:

* The ASR is not correct, but we are running an ASR pass. The contract is that if we are running an ASR pass, the ASR is valid both before and after.
* For performance reasons, the AST -> ASR must be just two passes at most. We need two passes due to first constructing a symbol table, and then using it, it's hard to do everything in single pass. But this change seems to add additional two passes. I think that is too much.

We should figure out the dependencies as we are constructing it, and the dependencies in declarations should be such that you only depend on what is specified previously in the code. If it is specified below the current line, then you can't use it in a declaration. That should make it possible to figure out the dependencies in the same pass that constructs the ASR.

Reply:
I see. 

> We should figure out the dependencies as we are constructing it, and the dependencies in declarations should be such that you only depend on what is specified previously in the code. 

I have thought about this and I can do this. However, it will be very bug prone and a lot tracking will be required to achieve this.

 What do you think of the following proposal?

"Merge this PR for now but as a follow up remove `PassUtils::UpdateDependenciesVisitor` from AST to ASR construction phase".

Let me know your thoughts. My main concern is that it would be very tricky to correctly track dependencies of several symbols in the AST to ASR phase. So it might take more time and efforts than usual. So I would suggest going for a slower but robust approach initially and then later on I will remove these `PassUtils::UpdateDependenciesVisitor` and make every thing work resulting in a faster solution. 

Also it depends on how urgent this feature is needed. Let me know about that as well.

Reply:
Well, this will be needed for LFortran as well then, and it will slow the compilation down for no good reason (I think).

So to go forward, we can extend ASR and implement whatever can be merged, but not to call this pass, that is, make it optional and off by default. Possibly exposed via a command line option.

I think we should not compromise on speed at all, as it is very hard to later recover, even if it seems simple.

Let's discuss a possible path forward at our next video meeting. If designed well, it seems to me it should be equivalent to the pass, so we maybe create some class or something to do this.